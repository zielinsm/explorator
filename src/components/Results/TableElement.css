.TableElement__pagination {
    margin-top: 0.5em !important;
    margin-bottom: 0.25em !important;
}
