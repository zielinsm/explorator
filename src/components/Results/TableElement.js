import React, { Component } from "react";
import arraySort from "array-sort";
import PropTypes from "prop-types";
import { Grid, Pagination, Table } from "semantic-ui-react";

import HeaderCell from "./HeaderCell";
import RowElement from "./RowElement";
import "./TableElement.css";

class TableElement extends Component {
  constructor(props) {
    super(props);
    this.data = props.results;
    this.perPage = props.perPage || 10;
    this.headerRow = props.headerRow;
    this.totalPages = Math.ceil(this.data.length / this.perPage);

    this.onSortChange = this.handleSortChange.bind(this);
    this.onRowClick = this.handleRowClick.bind(this);

    this.state = {
      data: this.data,
      activePage: 1,
      sort: "name",
      sortAscending: true
    };
  }

  pagination(data, activePage, perPage) {
    const finalElement = activePage * perPage;
    return data.slice(finalElement - perPage, finalElement);
  }

  sortData(data, column, sortAscending) {
    if (sortAscending) {
      return arraySort(data, column);
    } else {
      return arraySort(data, column, { reverse: true });
    }
  }

  handlePaginationChange = (e, { activePage }) => this.setState({ activePage });

  handleRowClick(e, fingerprint) {
    this.context.router.history.push("/node/" + fingerprint);
  }

  handleSortChange = (e, column) => {
    if (this.state.sort === column) {
      const opposite = !this.state.sortAscending;
      this.setState({ sortAscending: opposite });
    } else {
      this.setState({ sort: column });
    }
  };

  renderHeaderCell(data, index, handler) {
    return <HeaderCell cell={data} cellClick={handler} key={index} />;
  }

  renderBodyRow(data, handler) {
    return <RowElement key={data.fingerprint} row={data} rowClick={handler} />;
  }

  renderPagination(current, total, handler) {
    if (total > 1) {
      return (
        <Grid centered className="TableElement__pagination">
          <Pagination
            size="mini"
            activePage={current}
            totalPages={total}
            onPageChange={handler}
          />
        </Grid>
      );
    } else return null;
  }

  render() {
    const sorted = this.sortData(
      this.data,
      this.state.sort,
      this.state.sortAscending
    );
    const page = this.pagination(sorted, this.state.activePage, this.perPage);

    return (
      <div className="TableElement">
        <Table selectable>
          <Table.Header>
            <Table.Row>
              {this.headerRow.map((cell, index) =>
                this.renderHeaderCell(cell, index, this.onSortChange)
              )}
            </Table.Row>
          </Table.Header>
          <Table.Body>
            {page.map(row => this.renderBodyRow(row, this.onRowClick))}
          </Table.Body>
        </Table>
        {this.renderPagination(
          this.state.activePage,
          this.totalPages,
          this.handlePaginationChange
        )}
      </div>
    );
  }

  static contextTypes = {
    router: PropTypes.object
  };
}

export default TableElement;
