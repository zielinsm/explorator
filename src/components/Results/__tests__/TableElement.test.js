import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import TableElement from "../TableElement";

const mockResults = [
  {
    name: "mockNode",
    fingerprint: "MOCKNODE",
    address: "000.000.000.000",
    running: true
  },
  {
    name: "mockRelay",
    fingerprint: "MOCKRELAY",
    address: "000.000.000.000",
    running: true
  }
];

const headerRow = [
  { label: "Name", value: "name" },
  { label: "IP", value: "ip" },
  { label: "Fingerprint", value: "fingerprint" },
  { label: "Status", value: "running" }
];

describe("TableElement component:", () => {
  let wrapper;
  const component = (
    <TableElement results={mockResults} headerRow={headerRow} />
  );

  beforeEach(() => {
    wrapper = mount(component);
  });

  it("renders without crashing", () => {
    expect(wrapper.find(".TableElement")).toHaveLength(1);
  });

  it("sets the right amount of pages to render", () => {
    expect(wrapper.instance().totalPages).toEqual(1);
  });

  it("sets the right amount of pages to render", () => {
    expect(wrapper.instance().totalPages).toEqual(1);
  });

  it("responds to changes in sorting", () => {
    wrapper.instance().handleSortChange({}, "fingerprint");
    expect(wrapper.state("sort")).toEqual("fingerprint");
  });

  it("responds to changes in pagination", () => {
    wrapper.instance().handlePaginationChange({}, { activePage: 4 });
    expect(wrapper.state("activePage")).toEqual(4);
  });

  it("snapshot matches", () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
