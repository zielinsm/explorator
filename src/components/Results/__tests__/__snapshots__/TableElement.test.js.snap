// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableElement component: snapshot matches 1`] = `
<div
  className="TableElement"
>
  <table
    className="ui selectable table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className="HeaderCell"
          onClick={[Function]}
        >
          Name
        </th>
        <th
          className="HeaderCell"
          onClick={[Function]}
        >
          IP
        </th>
        <th
          className="HeaderCell"
          onClick={[Function]}
        >
          Fingerprint
        </th>
        <th
          className="HeaderCell"
          onClick={[Function]}
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className="RowElement"
        onClick={[Function]}
      >
        <td
          className=""
        >
          mockNode
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          MOCKNODE
        </td>
        <td
          className=""
        >
          <div
            className="ui green label"
            onClick={[Function]}
          >
            Running
          </div>
        </td>
      </tr>
      <tr
        className="RowElement"
        onClick={[Function]}
      >
        <td
          className=""
        >
          mockRelay
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          MOCKRELAY
        </td>
        <td
          className=""
        >
          <div
            className="ui green label"
            onClick={[Function]}
          >
            Running
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
